_posts/2021-07-13-JEV.md

```diff
@@ -0,0 +1,822 @@
+      // divisibility by 24 makes our lives easier 
+      assert src.length % 24 == 0;
+        
+      // init with min values
+      min[0] = min[1] = min[2] = Float.POSITIVE_INFINITY;
+      FloatVector firstMinVec   = FloatVector.broadcast(SPECIES, Float.POSITIVE_INFINITY);
+      FloatVector secondMinVec  = FloatVector.broadcast(SPECIES, Float.POSITIVE_INFINITY);
+      FloatVector thirdMinVec   = FloatVector.broadcast(SPECIES, Float.POSITIVE_INFINITY);
+      
+      // init with max values
+      max[0] = max[1] = max[2] = Float.NEGATIVE_INFINITY;
+      FloatVector firstMaxVec   = FloatVector.broadcast(SPECIES, Float.NEGATIVE_INFINITY);
+      FloatVector secondMaxVec  = FloatVector.broadcast(SPECIES, Float.NEGATIVE_INFINITY);
+      FloatVector thirdMaxVec   = FloatVector.broadcast(SPECIES, Float.NEGATIVE_INFINITY);  
+        
+      // increment by 24 floats * 32 bits = 768 bits
+      final int stride = 24;
+      for (int i = 0; i < src.length - stride; i += stride) {
+          // xyzxyzxy
+          final FloatVector firstVec = FloatVector.fromArray(SPECIES, src, i + 0);
+          firstMinVec = firstVec.min(firstMinVec);
+          firstMaxVec = firstVec.max(firstMaxVec);
+
+          // zxyzxyzx
+          final FloatVector secondVec = FloatVector.fromArray(SPECIES, src, i + 8);
+          secondMinVec = secondVec.min(secondMinVec);
+          secondMaxVec = secondVec.max(secondMaxVec);
+
+          // yzxyzxyz
+          final FloatVector thirdVec = FloatVector.fromArray(SPECIES, src, i + 16);
+          thirdMinVec = thirdVec.min(thirdMinVec);
+          thirdMaxVec = thirdVec.max(thirdMaxVec);
+        }
+        
+        // we now have 8 values in total for each of x, y, z
+        // so min/max-ing them takes a little more effort 
+      
+        final float[] allMins = new float[24];
+        firstMinVec.intoArray(allMins, 0);
+        secondMinVec.intoArray(allMins, 8);
+        thirdMinVec.intoArray(allMins, 16);
+
+        final float[] allMaxs = new float[24];
+        firstMaxVec.intoArray(allMaxs, 0);
+        secondMaxVec.intoArray(allMaxs, 8);
+        thirdMaxVec.intoArray(allMaxs, 16);
+
+        for (int i = 0; i < 8; i++) {
+            // x
+            min[0] = Math.min(min[0], allMins[(i * 3) + 0]);
+            max[0] = Math.max(max[0], allMaxs[(i * 3) + 0]);
+
+            // y
+            min[1] = Math.min(min[1], allMins[(i * 3) + 1]);
+            max[1] = Math.max(max[1], allMaxs[(i * 3) + 1]);
+
+            // z
+            min[2] = Math.min(min[2], allMins[(i * 3) + 2]);
+            max[2] = Math.max(max[2], allMaxs[(i * 3) + 2]);
+        }
+    }
+}
+```
+
+So, let's do a little benchmarking, and see how we did:
+
+|vanilla      |simd         |simd-768bit  |
+|---          |---          |---          |
+|34000ns      |18500ns      |6100ns       |
+
+Holy guacamole, that's almost 3 times as fast as the simd version and almost **6 times** faster than the vanilla one!!!!
```

Comment:
Apparently it does not unroll the previous to at least calculate on 18 floats each iteration using the multiple lanes?